File: scripts/scheduler.py
#!/usr/bin/env python3
"""Persistent, single-worker scheduler; no interactive Codex queue involved."""
import argparse
import datetime
import fcntl
import json
import os
from pathlib import Path
import signal
import subprocess
import tempfile
import time

ROOT = Path(os.environ.get('BOSS_AGENT_ROOT', Path(__file__).resolve().parents[1]))
MEMORY = ROOT / 'memory'
NODE_BINARY = os.environ.get('NODE_BINARY', 'node')


def atomic_json(path, value):
    fd, temporary = tempfile.mkstemp(dir=path.parent, prefix='.' + path.name)
    try:
        with os.fdopen(fd, 'w') as output:
            json.dump(value, output, ensure_ascii=False, indent=2)
            output.flush()
            os.fsync(output.fileno())
        os.replace(temporary, path)
    finally:
        if os.path.exists(temporary):
            os.unlink(temporary)


def timestamp():
    return datetime.datetime.now(datetime.timezone.utc).isoformat()


def run_worker(command, timeout=900):
    child = subprocess.Popen(command, cwd=ROOT, stdin=subprocess.DEVNULL,
                             start_new_session=True)
    try:
        return child.wait(timeout=timeout)
    except subprocess.TimeoutExpired:
        os.killpg(child.pid, signal.SIGTERM)
        try:
            child.wait(timeout=10)
        except subprocess.TimeoutExpired:
            os.killpg(child.pid, signal.SIGKILL)
            child.wait()
        return 124


def tick(force=False):
    with open(MEMORY / 'scheduler.lock', 'a') as lock:
        try:
            fcntl.flock(lock, fcntl.LOCK_EX | fcntl.LOCK_NB)
        except BlockingIOError:
            return 'already_running'
        if force:
            os.environ['JOB_AGENT_FORCE'] = '1'
        else:
            os.environ.pop('JOB_AGENT_FORCE', None)
        code = run_worker([NODE_BINARY, 'scripts/scheduled-agent.mjs'])
        if code == 124:
            subprocess.run([NODE_BINARY, 'scripts/harness-state.mjs', 'timeout'], cwd=ROOT, timeout=15, check=True)
            path = MEMORY / 'scheduled-cycle.json'
            cycle = json.loads(path.read_text()) if path.exists() else {}
            cycle.update(status='blocked', reason='worker_timeout_result_unconfirmed',
                         completedAt=timestamp())
            # Core cycle is already committed to Harness SQLite by harness-state.
            atomic_json(MEMORY / 'scheduler-status.json', {
                'checkedAt': timestamp(), 'state': 'blocked',
                'cycle': cycle.get('id'), 'reason': cycle['reason']})
        subprocess.run([NODE_BINARY, 'scripts/telegram-notify.mjs'], cwd=ROOT,
                       stdin=subprocess.DEVNULL, timeout=75, check=False)
        return code


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--once', action='store_true')
    args = parser.parse_args()
    os.umask(0o077)
    while True:
        started = time.monotonic()
        try:
            result = tick(force=args.once)
            print(json.dumps({'at': timestamp(), 'tickResult': result}), flush=True)
        except Exception as error:
            subprocess.run([NODE_BINARY, 'scripts/harness-state.mjs', 'error'], cwd=ROOT, timeout=15, check=False)
            atomic_json(MEMORY / 'scheduler-status.json', {
                'checkedAt': timestamp(), 'state': 'error', 'reason': str(error)})
            print(str(error), flush=True)
            result = 1
        if args.once:
            return 0 if result in (0, 'already_running') else 1
        # Poll health/notifications every minute; dispatcher persists the 30-min
        # outreach cadence and retry backoff. Restarts cannot bypass that cadence.
        atomic_json(MEMORY / 'scheduler-heartbeat.json', {'at': timestamp(), 'pid': os.getpid()})
        time.sleep(max(1, 60 - (time.monotonic() - started)))


if __name__ == '__main__':
    raise SystemExit(main())

File: src/runtime/scheduler.ts
import fs from "node:fs";
import path from "node:path";
import { randomUUID } from "node:crypto";
import { setTimeout as delay } from "node:timers/promises";
import { runProcessGroup } from "./process-supervisor.ts";

export interface SupervisorConfig {
  pollMs: number;
  workerTimeoutMs: number;
  terminationGraceMs: number;
  stateTimeoutMs: number;
  notificationTimeoutMs: number;
}
export function supervisorConfig(
  value: Partial<SupervisorConfig> = {},
): SupervisorConfig {
  const config = {
    pollMs: 60000,
    workerTimeoutMs: 900000,
    terminationGraceMs: 10000,
    stateTimeoutMs: 15000,
    notificationTimeoutMs: 75000,
    ...value,
  };
  if (
    !Object.values(config).every((v) => Number.isSafeInteger(v) && v > 0) ||
    config.workerTimeoutMs > 900000 ||
    config.terminationGraceMs > 30000
  )
    throw Error("监督器配置无效");
  return config;
}
export function writeDiagnostic(file: string, value: unknown): void {
  const temporary = path.join(
    path.dirname(file),
    "." + path.basename(file) + "." + randomUUID(),
  );
  const fd = fs.openSync(temporary, "wx", 0o600);
  try {
    fs.writeFileSync(fd, JSON.stringify(value));
    fs.fsyncSync(fd);
  } finally {
    fs.closeSync(fd);
  }
  fs.renameSync(temporary, file);
}
export async function schedulerTick(
  root: string,
  config: SupervisorConfig,
  force = false,
  signal?: AbortSignal,
  codeRoot = root,
): Promise<number | "already_running"> {
  const env: NodeJS.ProcessEnv = { ...process.env, BOSS_DATA_DIR: root };
  delete env.JOB_AGENT_FORCE;
  if (force) env.JOB_AGENT_FORCE = "1";
  const options = {
    cwd: root,
    env,
    signal,
    graceMs: config.terminationGraceMs,
  };
  const code = await runProcessGroup(
    "/usr/bin/flock",
    [
      "-n",
      "-E",
      "73",
      path.join(root, "memory/scheduler.lock"),
      process.execPath,
      path.join(codeRoot, "scripts/scheduled-agent.ts"),
    ],
    { ...options, timeoutMs: config.workerTimeoutMs },
  );
  if (code === 73) return "already_running";
  if (code === 130) return code;
  if (code === 124) {
    const committed = await runProcessGroup(
      process.execPath,
      [path.join(codeRoot, "scripts/harness-state.ts"), "timeout"],
      { ...options, timeoutMs: config.stateTimeoutMs },
    );
    if (committed !== 0) throw Error("超时状态提交失败");
  }
  await runProcessGroup(
    process.execPath,
    [path.join(codeRoot, "scripts/telegram-notify.ts")],
    { ...options, timeoutMs: config.notificationTimeoutMs },
  );
  return code;
}
export async function runScheduler(
  root: string,
  config: SupervisorConfig,
  once = false,
  signal?: AbortSignal,
  codeRoot = root,
): Promise<number> {
  process.umask(0o077);
  while (!signal?.aborted) {
    const started = Date.now();
    let result: number | "already_running" = 1;
    try {
      result = await schedulerTick(root, config, once, signal, codeRoot);
    } catch (error) {
      console.error(
        "调度错误：" + (error instanceof Error ? error.message : String(error)),
      );
      if (!signal?.aborted)
        try {
          await runProcessGroup(
            process.execPath,
            [path.join(codeRoot, "scripts/harness-state.ts"), "error"],
            {
              cwd: root,
              env: { ...process.env, BOSS_DATA_DIR: root },
              timeoutMs: config.stateTimeoutMs,
              graceMs: config.terminationGraceMs,
              signal,
            },
          );
        } catch (stateError) {
          console.error("状态错误记录失败：" + String(stateError));
        }
      writeDiagnostic(path.join(root, "memory/scheduler-health-error.json"), {
        checkedAt: new Date().toISOString(),
        state: "error",
        reason: error instanceof Error ? error.message : String(error),
      });
    }
    console.log(
      JSON.stringify({ at: new Date().toISOString(), tickResult: result }),
    );
    if (once) return result === 0 || result === "already_running" ? 0 : 1;
    writeDiagnostic(path.join(root, "memory/scheduler-heartbeat.json"), {
      at: new Date().toISOString(),
      pid: process.pid,
    });
    if (signal?.aborted) break;
    try {
      await delay(
        Math.max(1, config.pollMs - (Date.now() - started)),
        undefined,
        { signal },
      );
    } catch (error) {
      if (!signal?.aborted) throw error;
    }
  }
  return 0;
}
